skiko/src/awtMain/kotlin/org/jetbrains/skiko/FontInterop.awt.kt

```diff
@@ -0,0 +1,109 @@
+    @Volatile
+    private var allFontsCachedImpl = false
+
+    init {
+        GlobalScope.launch(Dispatchers.IO) {
+            cacheSystemFonts()
+            allFontsCachedImpl = true
+        }
+    }
+
+    private val systemFontNames: Array<String>
+        get() = GraphicsEnvironment.getLocalGraphicsEnvironment().availableFontFamilyNames
+
+    private val systemFontsPaths: Array<String> by lazy {
+        when (hostOs) {
+            OS.Windows -> {
+                val winPath = System.getenv("WINDIR")
+                val localAppPath = System.getenv("LOCALAPPDATA")
+                arrayOf(
+                    "$winPath\\Fonts",
+                    "$localAppPath\\Microsoft\\Windows\\Fonts"
+                )
+            }
+            OS.MacOS -> {
+                arrayOf(
+                    System.getProperty("user.home") + File.separator + "Library/Fonts",
+                    "/Library/Fonts",
+                    "/System/Library/Fonts"
+                )
+            }
+            OS.Linux -> {
+                val pathsToCheck = arrayOf(
+                    System.getProperty("user.home") + File.separator + ".fonts",
+                    "/usr/share/fonts/truetype",
+                    "/usr/share/fonts/TTF"
+                )
+                val resultList = ArrayList<String>()
+                for (i in pathsToCheck.indices.reversed()) {
+                    val path = pathsToCheck[i]
+                    val tmp = File(path)
+                    if (tmp.exists() && tmp.isDirectory && tmp.canRead()) {
+                        resultList.add(path)
+                    }
+                }
+                resultList.toTypedArray()
+            }
+            else -> {
+                throw RuntimeException("Unknown OS: $hostOs")
+            }
+        }
+    }
+
+    private val systemFontFiles: List<File> by lazy {
+        val extensions = arrayOf("ttf", "TTF")
+        val paths = systemFontsPaths
+        val files = mutableListOf<File>()
+        for (i in paths.indices) {
+            val fontDirectory = File(paths[i])
+            if (!fontDirectory.exists()) break
+            fontDirectory.walk().filter { it.isFile && it.extension in extensions }.forEach {
+                files.add(it)
+            }
+        }
+        files
+    }
+
+    private fun cacheSystemFonts() {
+        val fontFiles = systemFontFiles
+        for (file in fontFiles) {
+            try {
+                if (!fontsMap.containsValue(file.absoluteFile)) {
+                    val f = Font.createFont(Font.TRUETYPE_FONT, FileInputStream(file.absolutePath))
+                    val name = f.family
+                    fontsMap[name] = file.absoluteFile
+                }
+            } catch (e: FontFormatException) {
+            } catch (e: IOException) {
+            }
+        }
+    }
+
+    fun findFontFile(font: Font): File? {
+        return fontsMap[font.family]
+    }
+
+    val allFontsCached: Boolean
```

Comment:
Can we make caching as implementation details, so we can change it in the future?

Also, we can get rid of `Font` from API and receive only `familyName: String`. And maybe even get rid of AWT from API.

I suggest this API:
```
class SystemFonts {
    val familyNames: List<String>    

    fun fileOf(familyName: String): File

    companion object {
       suspend fun load(): SystemFonts
    }
}
```

It is no longer global, and requires manual creation/passing, but global operations/API's has its own flaws (we can discuss).

Reply:
> Can we make caching as implementation details, so we can change it in the future?

I can explain why this is bad.

Imagine a very big application. Some developer wants to know a file of some font and finds a class `AwtFontManager`, finds the method `findFontFile`, and starts using it. In 9 cases of 10 it works, because fonts are loaded in time. To fix 1 case when it doesn't work, it is needed to catch that case, investigate, and find out, that AwtFontManager requires awaiting for the loading.

I believe a good API should enforce it via its shape, not describing it via docs.

Reply:
I reworked API.

Reply:
> Can we make caching as implementation details, so we can change it in the future?
> 
> Also, we can get rid of `Font` from API and receive only `familyName: String`. And maybe even get rid of AWT from API.
> 
> I suggest this API:
> 
> ```
> class SystemFonts {
>     val familyNames: List<String>    
> 
>     fun fileOf(familyName: String): File
> 
>     companion object {
>        suspend fun load(): SystemFonts
>     }
> }
> ```
> 
> It is no longer global, and requires manual creation/passing, but global operations/API's has its own flaws (we can discuss).

Don't think like this API much.

Reply:
Because it has different purpose, which makes sense to exist in common code. This class is about AWT interop, and fact how we discover fonts (via system dirs) is an implementation detail.